internal/xds/rbac/rbac_engine.go

```diff
@@ -15,68 +15,210 @@
  */
 
 // Package rbac provides service-level and method-level access control for a
-// service.
+// service. See
+// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#role-based-access-control-rbac
+// for documentation.
 package rbac
 
 import (
+	"context"
+	"crypto/x509"
+	"errors"
 	"net"
+	"strconv"
 
 	v3rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
 	"google.golang.org/grpc/metadata"
 	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
 )
 
-// Engine is used for matching incoming RPCs to policies.
-type Engine struct {
+// ChainEngine represents a chain of RBAC Engines, used to make authorization
+// decisions on incoming RPCs.
+type ChainEngine struct {
+	chainedEngines []*engine
+}
+
+// NewChainEngine returns a chain of RBAC engines, used to make authorization
+// decisions on incoming RPCs. Returns a non-nil error for invalid policies.
+func NewChainEngine(policies []*v3rbacpb.RBAC) (*ChainEngine, error) {
+	var engines []*engine
+	for _, policy := range policies {
+		engine, err := newEngine(policy)
+		if err != nil {
+			return nil, err
+		}
+		engines = append(engines, engine)
+	}
+	return &ChainEngine{chainedEngines: engines}, nil
+}
+
+// IsAuthorized determines if an incoming RPC is authorized based on the chain of RBAC
+// engines and their associated actions.
+//
+// Errors returned by this function are compatible with the status package.
+func (cre *ChainEngine) IsAuthorized(ctx context.Context, methodName string) error {
```

Comment:
Is it required to pass the `methodName` as an argument to this method?

I see that there is a utility function named `Method` which return the method name given a context. See: https://github.com/grpc/grpc-go/blob/30dfb4b933a50fd366d7ed36ed4f71dbba2d382e/server.go#L1862

The only requirement is that the context wraps a server transport stream, i.e, it is created using [NewContextWithServerTransportStream](https://github.com/grpc/grpc-go/blob/30dfb4b933a50fd366d7ed36ed4f71dbba2d382e/server.go#L1670). I see that both `processUnaryRPC` and `processStreamingRPC` call this function to stash the server transport into the context.

Reply:
Hmmmm good point. Let me triage further.

Reply:
Yeah, great catch. Both NewContextWithServerTransportStream calls pipe transport into context, which then is passed into unary/streaming interceptors. With unary, context is explicitly passed, whereas in streaming, it becomes the server streams context. Thus, both types of interceptors will have a context with a transport stream present, and you can use this util function.

Reply:
Changed API to only have context, pulled methodName from context.